data_model/src/permission/validator.rs

```diff
@@ -0,0 +1,230 @@
+        }
+    }
+}
+
+/// Type of validator
+#[derive(
+    Debug,
+    Display,
+    Copy,
+    Clone,
+    PartialEq,
+    Eq,
+    Encode,
+    Decode,
+    Deserialize,
+    Serialize,
+    IntoSchema,
+    Hash,
+)]
+pub enum Type {
+    /// Validator checking [`Transaction`]
+    Transaction,
+    /// Validator checking [`Instruction`]
+    Instruction,
+    /// Validator checking [`QueryBox`]
+    Query,
+    /// Validator checking [`Expression`]
+    Expression,
+}
+
+/// Operation for which the permission should be checked
+pub trait NeedsPermission {
+    /// Get the type of validator required to check the operation
+    ///
+    /// Accepts `self` because of the [`NeedsPermissionBox`]
+    fn required_validator_type(&self) -> Type;
+}
+
+impl NeedsPermission for Instruction {
+    fn required_validator_type(&self) -> Type {
+        Type::Instruction
+    }
+}
+
+impl NeedsPermission for QueryBox {
+    fn required_validator_type(&self) -> Type {
+        Type::Query
+    }
+}
+
+// Expression might contain a query, therefore needs to be checked.
+impl NeedsPermission for Expression {
+    fn required_validator_type(&self) -> Type {
+        Type::Expression
+    }
+}
+
+/// Boxed version of [`NeedsPermission`]
+#[derive(
+    Debug,
+    Display,
+    Clone,
+    derive_more::From,
+    derive_more::TryInto,
+    Encode,
+    Decode,
+    Serialize,
+    Deserialize,
+)]
+pub enum NeedsPermissionBox {
+    /// [`Transaction`] application operation
+    Transaction(Transaction),
+    /// [`Instruction`] execution operation
+    Instruction(Instruction),
+    /// [`QueryBox`] execution operations
+    Query(QueryBox),
+    /// [`Expression`] evaluation operation
+    Expression(Expression),
+}
+
+impl NeedsPermission for NeedsPermissionBox {
+    fn required_validator_type(&self) -> Type {
+        match self {
+            NeedsPermissionBox::Transaction(_) => Type::Transaction,
+            NeedsPermissionBox::Instruction(_) => Type::Instruction,
+            NeedsPermissionBox::Query(_) => Type::Query,
+            NeedsPermissionBox::Expression(_) => Type::Expression,
+        }
+    }
+}
+
+/// Validation verdict. All *runtime validators* should return this type.
+///
+/// All operations are considered to be **valid** unless proven otherwise.
+/// Validators are allowed to either pass an operation to the next validator or to deny an operation.
+///
+/// # Note
+///
+/// There is no `Allow` variant because it has a wrong connotation and suggests an
+/// incorrect interpretation of validators system. All operations are allowed by default.
+/// Validators are checking for operation **incorrectness**, not for operation correctness.
+#[derive(Debug, Clone, PartialEq, Eq, Encode, Decode, Deserialize, Serialize, IntoSchema)]
+pub enum Verdict {
+    /// Operation is approved to pass to the next validator
+    /// or to be executed if there are no more validators
+    Pass,
+    /// Operation is denied
+    Deny(DenialReason),
+}
```

Comment:
The reason is the same as for "*Why `Pass` instead of `Allow`*".
`Allow` and `Result` have a wrong connotation and suggest an incorrect interpretation of validators system.
All operations are allowed by default. Validators are checking for operation **incorrectness**, not for operation correctness.

I'll write a note about `Result`.

It's my vision, but if you are not agree, we can discuss this